git-url-parse doesn't properly parse Bitbucket Server URIs

# Issue Type

  - Bug report

  # Gilt details

  ```
  gilt, version 1.2.2.dev2
  ```

  - Gilt installation method: source

  # Desired Behaviour

  Running `gilt overlay` using this `gilt.yml` config file:
  ```yaml
  - git: https://our.bitbucket.server/git/scm/myproject/my-helloworld-roles.git
    version: master
    files:
      - src: helloworld_*
        dst: roles/
  ```
  I expect everything matching `helloworld_.*` to be put in `./roles`

  # Actual Behaviour (Bug report only)

  ```pytb
  $ gilt --debug overlay
  Traceback (most recent call last):
    File "/home/philipsd6/.local/bin/gilt", line 10, in <module>
      sys.exit(main())
    File "/home/philipsd6/.local/lib/python3.6/site-packages/click/core.py", line 722, in __call__
      return self.main(*args, **kwargs)
    File "/home/philipsd6/.local/lib/python3.6/site-packages/click/core.py", line 697, in main
      rv = self.invoke(ctx)
    File "/home/philipsd6/.local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
      return _process_result(sub_ctx.command.invoke(sub_ctx))
    File "/home/philipsd6/.local/lib/python3.6/site-packages/click/core.py", line 895, in invoke
      return ctx.invoke(self.callback, **ctx.params)
    File "/home/philipsd6/.local/lib/python3.6/site-packages/click/core.py", line 535, in invoke
      return callback(*args, **kwargs)
    File "/home/philipsd6/.local/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
      return f(get_current_context(), *args, **kwargs)
    File "/home/philipsd6/devel/gilt/gilt/shell.py", line 68, in overlay
      with fasteners.InterProcessLock(c.lock_file):
    File "/home/philipsd6/.local/lib/python3.6/site-packages/fasteners/process_lock.py", line 174, in __enter__
      self.acquire()
    File "/home/philipsd6/.local/lib/python3.6/site-packages/fasteners/process_lock.py", line 151, in acquire
      self._do_open()
    File "/home/philipsd6/.local/lib/python3.6/site-packages/fasteners/process_lock.py", line 115, in _do_open
      made_basedir = _ensure_tree(basedir)
    File "/home/philipsd6/.local/lib/python3.6/site-packages/fasteners/process_lock.py", line 37, in _ensure_tree
      os.makedirs(path)
    File "/usr/lib/python3.6/os.py", line 210, in makedirs
      makedirs(head, mode, exist_ok)
    File "/usr/lib/python3.6/os.py", line 210, in makedirs
      makedirs(head, mode, exist_ok)
    File "/usr/lib/python3.6/os.py", line 220, in makedirs
      mkdir(name, mode)
  PermissionError: [Errno 13] Permission denied: '/our.bitbucket.server'
  ```

The problem is in the git-url-parse dependent package:
```pycon
In [1]: from giturlparse import parse
In [2]: p = parse('https://our.bitbucket.server/git/scm/myproject/my-helloworld-roles.git')
In [3]: assert p.name == 'my-helloworld-roles'
In [4]: assert p.owner == 'myproject', f"p.owner is incorrect: {p.owner}"
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-4-a20d961cd7d8> in <module>
----> 1 assert p.owner == 'myproject', f"p.owner is incorrect: {p.owner}"

AssertionError: p.owner is incorrect: /our.bitbucket.server/git/scm/myproject

In [5]: assert p.resource == 'our.bitbucket.server', f"p.resource is incorrect: {p.resource}"
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-5-88ee6374fc22> in <module>
----> 1 assert p.resource == 'our.bitbucket.server', f"p.resource is incorrect: {p.resource}"

AssertionError: p.resource is incorrect: https

In [6]: assert p.protocol == 'https', f"p.protocol is incorrect: {p.protocol}"
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-12-86dec67945be> in <module>
----> 1 assert p.protocol == 'https', f"p.protocol is incorrect: {p.protocol}"

AssertionError: p.protocol is incorrect: ssh
```

--- a/gilt/config.py
+++ b/gilt/config.py
@@ -24,10 +24,15 @@ import collections
 import errno
 import os
 
-import giturlparse
 import yaml
 
 from gilt import interpolation
+
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    # Python 2
+    from urlparse import urlparse
 
 
 class ParseError(Exception):
@@ -59,6 +64,37 @@ def config(filename):
     return [Config(**d) for d in _get_config_generator(filename)]
 
 
+def _parse_repo_uri(uri, scm="git"):
+    """
+    Construct and return a `ParsedRepo` object.
+
+    :param uri: A SCM repository URI.
+    :return: ParsedRepo
+    """
+    ParsedRepo = collections.namedtuple('ParsedRepo', [
+        'hostname',
+        'owner',
+        'name'
+    ])
+
+    scm_ext = "." + scm
+    o = urlparse(uri)
+    if not o.hostname:  # scp-style "URI", so fake it
+        uri = "ssh://" + uri.replace(":/", "/").replace(":", "/")
+        o = urlparse(uri)
+    path = o.path
+    if path.endswith(scm_ext):
+        path = path[0:-len(scm_ext)]
+    try:
+        name, owner = path.rsplit("/", 2)[-1:-3:-1]
+        owner = owner.lstrip("~")
+    except ValueError:
+        # if repo is the root of the path, then there is no owner
+        owner = ''
+
+    return ParsedRepo(o.hostname, owner, name)
+
+
 def _get_files_config(src_dir, files_list):
     """
     Construct `FileConfig` object and return a list.
@@ -85,9 +121,12 @@ def _get_config_generator(filename):
     """
     for d in _get_config(filename):
         repo = d['git']
-        parsedrepo = giturlparse.parse(repo)
-        name = '{}.{}'.format(parsedrepo.owner, parsedrepo.name)
-        src_dir = os.path.join(_get_clone_dir(), name)
+        parsedrepo = _parse_repo_uri(repo)
+        if parsedrepo.owner:
+            name = '{}.{}'.format(parsedrepo.owner, parsedrepo.name)
+        else:
+            name = parsedrepo.name
+        src_dir = os.path.join(_get_clone_dir(), parsedrepo.hostname, name)
         files = d.get('files')
         post_commands = d.get('post_commands', [])
         dst_dir = None
@@ -95,7 +134,9 @@ def _get_config_generator(filename):
             dst_dir = _get_dst_dir(d['dst'])
         yield {
             'git': repo,
-            'lock_file': _get_lock_file(name),
+            'lock_file': os.path.join(
+                _get_lock_dir(), parsedrepo.hostname, name
+            ),
             'version': d['version'],
             'name': name,
             'src': src_dir,
@@ -154,13 +195,6 @@ def _get_dst_dir(dst_dir):
     _makedirs(dst_dir)
 
     return os.path.join(wd, dst_dir)
-
-
-def _get_lock_file(name):
-    """ Return the lock file for the given name. """
-    return os.path.join(
-        _get_lock_dir(),
-        name, )
 
 
 def _get_base_dir():

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,7 +1,6 @@
 click
 colorama
 fasteners
-git-url-parse
 pbr
 PyYAML
 sh

--- a/test/test_config.py
+++ b/test/test_config.py
@@ -59,6 +59,170 @@ def test_config(gilt_config_file):
     assert ('library', '') == os_split(f.dst)[-2:]
 
 
+gilt_repos = [
+    (
+        "example.com:/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "example.com:owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "example.com:repo.git",
+        {"name": "repo", "owner": "", "hostname": "example.com"},
+    ),
+    (
+        "git+https://example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "git+ssh://example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "git+ssh://git@git.example.com/~philip.sd6/test.repo.git",
+        {
+            "name": "test.repo",
+            "owner": "philip.sd6",
+            "hostname": "git.example.com",
+        },
+    ),
+    (
+        "git://example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "http://example.com/owner/repo",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "http://example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "http://example.com/repo",
+        {"name": "repo", "owner": "", "hostname": "example.com"},
+    ),
+    (
+        "http://example.com:29418/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "http://user@example.com/user/repo",
+        {"name": "repo", "owner": "user", "hostname": "example.com"},
+    ),
+    (
+        "http://user@example.com:29418/user/repo",
+        {"name": "repo", "owner": "user", "hostname": "example.com"},
+    ),
+    (
+        "https://example.com/git/scm/project/my-sample.repo.git",
+        {
+            "name": "my-sample.repo",
+            "owner": "project",
+            "hostname": "example.com",
+        },
+    ),
+    (
+        "https://example.com/git/scm/~philip.sd6/my-sample.repo.git",
+        {
+            "name": "my-sample.repo",
+            "owner": "philip.sd6",
+            "hostname": "example.com",
+        },
+    ),
+    (
+        "https://example.com/owner/repo",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "https://example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "https://example.com/repo",
+        {"name": "repo", "owner": "", "hostname": "example.com"},
+    ),
+    (
+        "https://example.com:29418/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "https://github.com/sphinx-doc/sphinx.git",
+        {"name": "sphinx", "owner": "sphinx-doc", "hostname": "github.com"},
+    ),
+    (
+        "https://github.com/tterranigma/Stouts.openvpn",
+        {
+            "name": "Stouts.openvpn",
+            "owner": "tterranigma",
+            "hostname": "github.com",
+        },
+    ),
+    (
+        "https://github.com/tterranigma/Stouts.openvpn.git",
+        {
+            "name": "Stouts.openvpn",
+            "owner": "tterranigma",
+            "hostname": "github.com",
+        },
+    ),
+    (
+        "https://user@example.com/user/repo",
+        {"name": "repo", "owner": "user", "hostname": "example.com"},
+    ),
+    (
+        "https://user@example.com:29418/user/repo",
+        {"name": "repo", "owner": "user", "hostname": "example.com"},
+    ),
+    (
+        "rsync://example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "ssh://example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "ssh://example.com:29418/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "ssh://user@example.com/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "ssh://user@example.com:29418/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "user@example.com:/owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "user@example.com:owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "example.com"},
+    ),
+    (
+        "user@example.com:repo.git",
+        {"name": "repo", "owner": "", "hostname": "example.com"},
+    ),
+    (
+        "user@foo-example.com:owner/repo.git",
+        {"name": "repo", "owner": "owner", "hostname": "foo-example.com"},
+    ),
+]
+
+
+@pytest.mark.parametrize('uri,expected', gilt_repos)
+def test_config_repo(uri, expected):
+    parsedrepo = config._parse_repo_uri(uri)
+    assert parsedrepo.hostname == expected['hostname']
+    assert parsedrepo.owner == expected['owner']
+    assert parsedrepo.name == expected['name']
+
+
 @pytest.fixture()
 def missing_git_key_data():
     return [{'foo': 'https://github.com/retr0h/ansible-etcd.git'}]

--- a/test/test_config.py
+++ b/test/test_config.py
@@ -38,9 +38,10 @@ def test_config(gilt_config_file):
     assert 'https://github.com/retr0h/ansible-etcd.git' == r.git
     assert 'master' == r.version
     assert 'retr0h.ansible-etcd' == r.name
-    assert (gilt_root, 'clone', 'retr0h.ansible-etcd') == os_split(r.src)[-3:]
-    assert (gilt_root, 'lock',
-            'retr0h.ansible-etcd') == os_split(r.lock_file)[-3:]
+    assert (gilt_root, 'clone', 'github.com',
+            'retr0h.ansible-etcd') == os_split(r.src)[-4:]
+    assert (gilt_root, 'lock', 'github.com',
+            'retr0h.ansible-etcd') == os_split(r.lock_file)[-4:]
     assert ('roles', 'retr0h.ansible-etcd', '') == os_split(r.dst)[-3:]
     assert [] == r.files
 
@@ -52,8 +53,9 @@ def test_config(gilt_config_file):
     assert r.dst is None
 
     f = r.files[0]
-    x = (gilt_root, 'clone', 'lorin.openstack-ansible-modules', '*_manage')
-    assert x == os_split(f.src)[-4:]
+    x = (gilt_root, 'clone', 'github.com',
+         'lorin.openstack-ansible-modules', '*_manage')
+    assert x == os_split(f.src)[-5:]
     assert ('library', '') == os_split(f.dst)[-2:]
 
 

